Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+        return;
+    case FPU_Exception::Underflow:
+        m_fpu_error_underflow = 1;
+        if (!m_fpu_mask_underflow)
+            break;
+        return;
+    case FPU_Exception::Precision:
+        m_fpu_error_precision = 1;
+        if (!m_fpu_mask_precision)
+            break;
+        return;
+    }
+
+    // set exception bit
+    m_fpu_error_summary = 1;
+
+    // FIXME: set traceback
+    // For that we need to get the currently executing instruction and
+    // the previous eip
+
+    // FIXME: Call FPU Exception handler
+    reportln("Trying to call Exception handler from {}", fpu_exception_string(ex));
+    fpu_dump_env();
+    m_emulator.dump_backtrace();
+    TODO();
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round(long double value) const
+{
+    // FIXME: may need to set indefinite values manually
+    switch (fpu_get_round_mode()) {
+    case RoundingMode::NEAREST:
+        return static_cast<T>(roundl(value));
+    case RoundingMode::DOWN:
+        return static_cast<T>(floorl(value));
+    case RoundingMode::UP:
+        return static_cast<T>(ceill(value));
+    case RoundingMode::TRUNK:
+        return static_cast<T>(truncl(value));
+    default:
+        VERIFY_NOT_REACHED();
+    }
+}
+
+template<Arithmetic T>
+ALWAYS_INLINE T SoftFPU::fpu_round_checked(long double value)
+{
+    T result = fpu_round<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert(long double value) const
+{
+    // FIXME: actually round the right way
+    return static_cast<T>(value);
+}
+template<FloatingPoint T>
+ALWAYS_INLINE T SoftFPU::fpu_convert_checked(long double value)
+{
+    T result = fpu_convert<T>(value);
+    if (auto rnd = value - result) {
+        if (rnd > 0)
+            set_c1(1);
+        else
+            set_c1(0);
+        fpu_set_exception(FPU_Exception::Precision);
+    }
+    return result;
+}
+
+// Instructions
+
+// DATA TRANSFER
+void SoftFPU::FLD_RM32(const X86::Instruction& insn)
+{
+    if (insn.modrm().is_register()) {
+        fpu_push(fpu_get(insn.modrm().register_index()));
+    } else {
+        auto new_f32 = insn.modrm().read32(m_cpu, insn);
+        // FIXME: Respect shadow values
+        fpu_push(bit_cast<float>(new_f32.value()));
+    }
+}
+void SoftFPU::FLD_RM64(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto new_f64 = insn.modrm().read64(m_cpu, insn);
+    // FIXME: Respect shadow values
+    fpu_push(bit_cast<double>(new_f64.value()));
+}
+void SoftFPU::FLD_RM80(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+
+    // long doubles can be up to 128 bits wide in memory for reasons (alignment) and only uses 80 bits of precision
+    // GCC uses 12 bytes in 32 bit and 16 bytes in 64 bit mode
```

Comment:
GCC has flags to select 10, 12 or 16 byte size in 32-bit mode.

Reply:
I did not know that, and could not find anything about it,
you are free to make a pr for that

Reply:
-m128bit-long-double and -m96bit-long-double do this. Turns out I misremembered a 10-bit version existing.

Reply:
sound like it might actually set long double to __float128, which needs another extension to work

Reply:
never mind, that would be -mlong-double(64|80|128)